core/src/entry_verification_service.rs

```diff
@@ -0,0 +1,162 @@
+//! [`EntryVerificationService`] is responsible for running different verifications (signature, PoH)
+//! on [`Entry`]es as they are received from the network.  Before they are assembled into a
+//! [`Bank`].
+
+use {
+    self::{
+        error::{AddDataSetError, CleanupError, StartError, WaitForResultError},
+        shared_wrapper::SharedWrapper,
+        worker::Command,
+    },
+    crossbeam_channel,
+    derive_more::{Display, From, Into},
+    solana_entry::entry::Entry,
+    solana_ledger::blockstore::Blockstore,
+    solana_sdk::clock::Slot,
+    std::{
+        cmp::{Ord, PartialOrd},
+        collections::HashMap,
+        ops::RangeInclusive,
+        result::Result,
+        sync::Arc,
+        time::Instant,
+    },
+};
+
+#[macro_use]
+pub mod error;
+mod shared_wrapper;
+mod worker;
+
+pub struct EntryVerificationService {
+    commands: crossbeam_channel::Sender<Command>,
+    results: Arc<SharedResults>,
+}
+
+impl EntryVerificationService {
+    pub fn start(blockstore: Arc<Blockstore>) -> Result<Self, StartError> {
+        let results = SharedResults::new(Results::default());
+
+        let commands = worker::start(blockstore, results.clone())?;
+        Ok(Self { commands, results })
+    }
+
+    /// Provides a new data set, including it in the verification process.
+    pub fn add_data_set(
+        &self,
+        slot: Slot,
+        shard_indices: RangeInclusive<u32>,
+        last_in_slot: bool,
+        entries: &[Entry],
+    ) -> Result<(), AddDataSetError> {
+        self.commands
+            .send(Command::AddDataSet {
+                slot,
+                shard_indices,
+                last_in_slot,
+                entries: Arc::new(entries.iter().cloned().collect()),
+            })
+            .map_err(|_| AddDataSetError::WorkerTerminated)
+    }
+
+    pub fn wait_for_result(
+        &self,
+        deadline: Instant,
+        slot: Slot,
+        data_set: DataSetFirstShred,
+    ) -> Result<bool, WaitForResultError> {
+        self.commands
+            .send(Command::WaitingFor { slot, data_set })
+            .map_err(|_| WaitForResultError::WorkerTerminated)?;
+
+        let wait_start = Instant::now();
+        match self.results.check_and_wait_until(deadline, |results| {
+            if slot < results.starting_at {
```

Comment:
Could use a comment here explaining what this means. I'm assuming we're waiting on a slot that has already been processed or something

Reply:
There is an explanation in the doc-comment for `Result`:
```rust
/// Results that are shared between the worker and the API of the service.
/// Wrapping it in [`SharedWrapper`] allows the API to wait for the worker, without actively
/// checking for the state to be updated.
struct Results {
    /// First slot that we are going to process and, possibly, have any results for.
    ///
    /// `EntryVerificationService::cleanup()` shifts this value forward.  And all the other
    /// operations are only valid, when called for this or later slots.
    starting_at: Slot,

    slots: HashMap<Slot, SlotVerificationStatus>,
}
```

I try to document the data structures first.
It is accessible in the LSP hints, though, you do not get that in GitHub %)
I think `Results::starting_at` doc-comment explains it.
But let me know if you think this deserves an additional explanation.

For the `set_slot_status()` case, ignoring an update for a slot that is before `starging_at` is the only choice.  As the data structure invariant is that we should not have any information in `slots`, for slots before `starting_at`.

Returning an error is an option, in general.  But in this particular case, it does not matter to any of the callers.  As it would be just a race with the cleanup, if everything else is working correctly.  I do not think I can avoid this race.  Though, I have not though about it too much.